### PR TITLE
Fix partial templates

### LINKED ISSUE #1779: Fix bug with partial templates
## Problem description
Pyccel fails during the semantic stage when trying to accelerate a function which uses a partial template `'T'`, if this not used alone for any of the arguments. That is, Pyccel works correctly if the arguments have types `('T', 'T[:]', 'T[:,:]')`, but fails if they have types `('T[:]', 'T[:,:]')`.

## To reproduce
In this code snippet
``` python
from pyccel.decorators import template

@template(name='T', types=[float, complex])
def mul_sum(a: 'T', v: 'T[:]'):
    return a * v.sum()

@template(name='T', types=[float, complex])
def just_sum(v: 'T[:]'):
    return v.sum()
```
the function `mul_sum` is pyccelized correctly, while the function `just_sum` leads to the error below.

## Error details
Provide the generated code, or the error message:
```bash/fortran/c
ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: test_template.py [8,13]| Undefined variable (T)
```

## Expected behavior
Pyccel should be able to process function `just_sum` with no issues. 

## Language
The error happens at the semantic stage, and it is independent of the language.

### PR DESCRIPTION
**Pull request by EmilyBourne**
Fix bug with partial templates (fixes #1779 ) where the partial template is incorrectly stripped away with unused template arguments. Add a test and update patch version.

### CODE REVIEW

File: tests/epyccel/test_epyccel_decorators.py [hunk 1]
```diff
@@ -146,3 +146,28 @@ def get_val(x : int = None , y : int = None):
...
+    def my_sum(v: 'T[:]'):
+        return v.sum()
+
+    pyccel_sum = epyccel(my_sum, language=language)
+
+    x = np.ones(4, dtype=float)
+    y = np.ones(4, dtype=float)
+
+    python_fl = my_sum(x)
+    pyccel_fl = pyccel_sum(y)
```
Comment (yguclu) [inline tests/epyccel/test_epyccel_decorators.py:161]: Why are two arrays needed?

File: tests/epyccel/test_epyccel_decorators.py [hunk 2]
```diff
@@ -146,3 +146,28 @@ def get_val(x : int = None , y : int = None):
...
+
+    assert python_fl == pyccel_fl
+    assert isinstance(python_fl, type(pyccel_fl))
+
+    x = np.ones(4, dtype=complex)
+    y = np.ones(4, dtype=complex)
```
Comment (yguclu) [inline tests/epyccel/test_epyccel_decorators.py:167]: Maybe we should have a proper complex test?
```suggestion
    x = np.ones(4, dtype=complex) * (1 + 3j)
    y = np.ones(4, dtype=complex)
```

File: tests/epyccel/test_epyccel_decorators.py (hunk 1 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_decorators.py:161]: They're not :+1:

File: tests/epyccel/test_epyccel_decorators.py (hunk 2 above)
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_decorators.py]: Fixed f0cd8cc6

Comment (yguclu) [inline tests/epyccel/test_epyccel_decorators.py]: Or better:
```suggestion
    x = np.full(4, 1 + 3j)
```

Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_decorators.py]: Fixed 201ec8b0

**Pull request merged**